### PR TITLE
Creating comma temps differently for SubMulDiv morph

### PR DESCRIPTION
**Pull request by TIHan**
**Description**

This changes how comma temps are created for SubMulDiv morph.

Example code:
```fsharp
(multiplier * seed2 + increment) % Modulus;
```

In morph, this would turn into:
```fsharp
(let tmp1 = multiplier * seed2 + increment) in tmp1) - (tmp1 / Modulus) * Modulus
 ```
 
This PR morphs it into:
```fsharp
let tmp1 = multiplier * seed2 + increment in
tmp1 - (tmp1 / Modulus) * Modulus
```

Basically, we are moving the comma temps to be the root of the expression instead of leafs.

**Acceptance Criteria**

- [x] CI passes
- [x] Investigate a possible generic way of lifting GT_COMMAs up to be root nodes instead of leaf nodes.
    - This might be a bit harder to do in a generic way. For now, let's not worry about it.
    
[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1793637&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by TIHan** [comment]
@dotnet/jit-contrib This is ready. The diffs will show some regressions, but they are not detrimental. The improvements on both x64 and ARM64 look good, especially ARM64.

**Comment by TIHan** [comment]
@kunalspathak @AndyAyersMS - I feel good about this change and it's ready. It should help optimizing 'mod' since I'll be able to do the transformation before rationalization.

**Comment by TIHan** [comment]
@kunalspathak @AndyAyersMS - I feel good about this change and it's ready. It should help optimizing 'mod' since I'll be able to do the transformation before rationalization.

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -1865,26 +1919,16 @@ GenTree* Compiler::fgInsertCommaFormTemp(GenTree** ppTree, CORINFO_CLASS_HANDLE
...
-        lvaSetStruct(lclNum, structType, false);
-    }
-
     // If subTree->TypeGet() == TYP_STRUCT, gtNewTempAssign() will create a GT_COPYBLK tree.
     // The type of GT_COPYBLK is TYP_VOID.  Therefore, we should use subTree->TypeGet() for
     // setting type of lcl vars created.
-    GenTree* asg = gtNewTempAssign(lclNum, subTree);
-
-    GenTree* load = new (this, GT_LCL_VAR) GenTreeLclVar(GT_LCL_VAR, subTree->TypeGet(), lclNum);
+    TempInfo tempInfo = fgMakeTemp(subTree, structType);
```
Comment (AndyAyersMS) [inline src/coreclr/jit/morph.cpp:1925]: Comment two lines up is now stale, should refer to `fgMakeTemp`.

File: src/coreclr/jit/morph.cpp [hunk 2]
```diff
@@ -13826,24 +13870,49 @@ GenTree* Compiler::fgMorphModToSubMulDiv(GenTreeOp* tree)
...
+    var_types type = div->gtType;
+
+    div->gtOp1 = gtClone(dividend);
+    div->gtOp2 = gtClone(divisor);
+
+    GenTree* const mul = gtNewOperNode(GT_MUL, type, div, divisor);
+    GenTree* const sub = gtNewOperNode(GT_SUB, type, dividend, mul);
+
+    GenTree* result = sub;
+    for (int i = tempInfoCount - 1; i >= 0; i--)
```
Comment (AndyAyersMS) [inline src/coreclr/jit/morph.cpp:13904]: Seems like you could just create the temps in the opposite order and avoid having to walk backwards through the array here.

Or if you think it's better this way, add a comment indicating why this ordering is needed.

File: src/coreclr/jit/morph.cpp [hunk 3]
```diff
@@ -13826,24 +13870,49 @@ GenTree* Compiler::fgMorphModToSubMulDiv(GenTreeOp* tree)
         noway_assert(!"Illegal gtOper in fgMorphModToSubMulDiv");
     }
 
-    var_types type = tree->gtType;
+    GenTreeOp* const div = tree;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/morph.cpp:13873]: Might be nice to put a before/after IR picture up in the method comment notes section.

Comment (TIHan) [reply src/coreclr/jit/morph.cpp:13895]: Done.

File: src/coreclr/jit/morph.cpp (hunk 2 above)
Comment (TIHan) [reply src/coreclr/jit/morph.cpp:13928]: IMO, I think it's better to walk backwards in the array than creating the temps in opposite order. I added a comment though describing why.

File: src/coreclr/jit/morph.cpp (hunk 1 above)
Comment (TIHan) [reply src/coreclr/jit/morph.cpp:1925]: Fixed. I moved the comment to inside of `fgMakeTemp` and modified it a tiny bit.

**Review by AndyAyersMS — APPROVED** [review]
Looks good.

(but: you have formatting to fix).

**Comment by JulieLeeMSFT** [comment]
> The diffs will show some regressions, but they are not detrimental. The improvements on both x64 and ARM64 look good, especially ARM64.

@TIHan can you show the perf results in this PR?

**Comment by TIHan** [comment]
Here is the link to the diffs: [Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1793637&view=ms.vss-build-web.run-extensions-tab)

File: src/coreclr/jit/morph.cpp [hunk 4]
```diff
@@ -13833,6 +13877,28 @@ GenTree* Compiler::fgMorphMultiOp(GenTreeMultiOp* multiOp)
 //    division will be used, in that case this transform allows CSE to
 //    eliminate the redundant div from code like "x = a / 3; y = a % 3;".
 //
+//    Before:
+//        *  RETURN    int
```
Comment (kunalspathak) [inline src/coreclr/jit/morph.cpp:13881]: Something doesn't feel right in `Before:` tree. Did it mean to have `(V00 * V00) % V01`. Same in `After:`.

**Comment by TIHan** [comment]
> Seems there is significant regression introduced on arm64:

It does look significant, but when I looked at the top regressions, the actual code is really wild with a lot of casts and checked contexts - something you really would not see in user code.

**Comment by TIHan** [comment]
> Seems like we are not CSEing the % operation for this:

That is a good catch. I'm surprised by this one - I'm even more surprised how it was able to CSE it before because one of the ops would have a `GT_COMMA` in it.

File: src/coreclr/jit/morph.cpp (hunk 4 above)
Comment (kunalspathak) [reply src/coreclr/jit/morph.cpp:13881]: As spoke offline, change `MUL` to `BINOP` and add a note that it is applicable for any `BINOP` that is safe to clone.

**Review by kunalspathak — APPROVED** [review]
LGTM

**Comment by TIHan** [comment]
As a note from @kunalspathak 's comment regarding the `%` CSE in this image:
https://user-images.githubusercontent.com/12488060/171420114-f42a8c64-54ca-476f-a9c8-43b13c6775fe.png
It is actually reversed; the left side is the `diff` whereas the right side was the `base`. This means that the change actually makes CSE work in this example :)

**Pull request merged**